{
 "metadata": {
  "name": "",
  "signature": "sha256:e5fa000cd52d5404b2532eac256402f6590fbe546ac5c7b01696137594c7765e"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%matplotlib inline"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 2
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from collections import defaultdict"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 3
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "This notebook details the procedure used to map the reads from the infection experiments to the database of short fragments, used for the mutations.\n",
      "To reproduce the analysis, some of the requirements are:\n",
      "\n",
      "- Python libraries used in this notebook\n",
      "- Bowtie2 installed in the path. You can get it from here http://bowtie-bio.sourceforge.net/bowtie2/index.shtml\n",
      "- Download the raw sequences from () and put in a folder called seqs.\n",
      "\n"
     ]
    },
    {
     "cell_type": "heading",
     "level": 3,
     "metadata": {},
     "source": [
      "Create the reference database"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "The original xls table was exported as a plain text table. The format of the table is:\n",
      "\n",
      "| Running No | Target gene| gene in LT2 | description | Primer 1 | Primer 2|\n",
      "|------|\n",
      "\n",
      "For the unique id of each gene, the first column Running No, will be used"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#Import file, and create a fasta file for Primer1 and Primer 2\n",
      "#The table is GeneSeqsTable.txt\n",
      "!mkdir GeneSeqs_DB\n",
      "\n",
      "#I will store the info of this table in a hash\n",
      "#This will be used to later add the information for the outputs\n",
      "\n",
      "table_dict_info = defaultdict(list)\n",
      "\n",
      "with open(\"GeneSeqTable.txt\") as fp:\n",
      "    for line in fp:\n",
      "        line = line.rstrip()\n",
      "        if line.strip():\n",
      "            running_no, target_gene, gene_LT2, desc, primer1, primer2 = line.split(\"\\t\")\n",
      "            \n",
      "            #I will use the running no as the ID to store in the dictionary. I need to check also\n",
      "            #that is not present\n",
      "            \n",
      "            if running_no in table_dict_info.keys():\n",
      "                print line \n",
      "                print \"Already there\"\n",
      "                break\n",
      "            else:\n",
      "                table_dict_info[running_no] = [target_gene, gene_LT2, desc, primer1, primer2]\n",
      "            "
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "mkdir: GeneSeqs_DB: File exists\r\n"
       ]
      }
     ],
     "prompt_number": 8
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#Using the dictionary in memory, now create the output for primer 1 and primer 2\n",
      "\n",
      "#Open the files\n",
      "primer1_file = open(\"GeneSeqs_DB/primer1.fna\", 'w')\n",
      "primer2_file = open(\"GeneSeqs_DB/primer2.fna\", 'w')\n",
      "\n",
      "for entry in table_dict_info:\n",
      "    primer1_seq = table_dict_info[entry][3]\n",
      "    primer2_seq = table_dict_info[entry][4]\n",
      "    \n",
      "    primer1_file.write(\">\" + entry + \"\\n\")\n",
      "    primer2_file.write(\">\" + entry + \"\\n\")\n",
      "    \n",
      "    primer1_file.write(primer1_seq + \"\\n\")\n",
      "    primer2_file.write(primer2_seq + \"\\n\")\n",
      "\n",
      "primer1_file.close()\n",
      "primer2_file.close()\n",
      "    \n",
      "    "
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 9
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#Index database using Bowtie2\n",
      "\n",
      "!bowtie2-build GeneSeqs_DB/primer1.fna GeneSeqs_DB/P1\n",
      "!bowtie2-build GeneSeqs_DB/primer2.fna GeneSeqs_DB/P2\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "Settings:\r\n",
        "  Output files: \"GeneSeqs_DB/P1.*.bt2\"\r\n",
        "  Line rate: 6 (line is 64 bytes)\r\n",
        "  Lines per side: 1 (side is 64 bytes)\r\n",
        "  Offset rate: 4 (one in 16)\r\n",
        "  FTable chars: 10\r\n",
        "  Strings: unpacked\r\n",
        "  Max bucket size: default\r\n",
        "  Max bucket size, sqrt multiplier: default\r\n",
        "  Max bucket size, len divisor: 4\r\n",
        "  Difference-cover sample period: 1024\r\n",
        "  Endianness: little\r\n",
        "  Actual local endianness: little\r\n",
        "  Sanity checking: disabled\r\n",
        "  Assertions: disabled\r\n",
        "  Random seed: 0\r\n",
        "  Sizeofs: void*:8, int:4, long:8, size_t:8\r\n",
        "Input files DNA, FASTA:\r\n",
        "  GeneSeqs_DB/primer1.fna\r\n",
        "Building a SMALL index\r\n",
        "Reading reference sizes\r\n",
        "Warning: Encountered reference sequence with only gaps\r\n"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "  Time reading reference sizes: 00:00:00\r\n",
        "Calculating joined length\r\n",
        "Writing header\r\n",
        "Reserving space for joined string\r\n",
        "Joining reference sequences\r\n",
        "  Time to join reference sequences: 00:00:00\r\n",
        "bmax according to bmaxDivN setting: 62535\r\n",
        "Using parameters --bmax 46902 --dcv 1024\r\n",
        "  Doing ahead-of-time memory usage test\r\n",
        "  Passed!  Constructing with these parameters: --bmax 46902 --dcv 1024\r\n",
        "Constructing suffix-array element generator\r\n",
        "Building DifferenceCoverSample\r\n",
        "  Building sPrime\r\n",
        "  Building sPrimeOrder\r\n",
        "  V-Sorting samples\r\n",
        "  V-Sorting samples time: 00:00:00\r\n",
        "  Allocating rank array\r\n",
        "  Ranking v-sort output\r\n",
        "  Ranking v-sort output time: 00:00:00\r\n",
        "  Invoking Larsson-Sadakane on ranks\r\n",
        "  Invoking Larsson-Sadakane on ranks time: 00:00:00\r\n",
        "  Sanity-checking and returning\r\n",
        "Building samples\r\n",
        "Reserving space for 12 sample suffixes\r\n",
        "Generating random suffixes\r\n",
        "QSorting 12 sample offsets, eliminating duplicates\r\n",
        "QSorting sample offsets, eliminating duplicates time: 00:00:00\r\n",
        "Multikey QSorting 12 samples\r\n",
        "  (Using difference cover)\r\n",
        "  Multikey QSorting samples time: 00:00:00\r\n",
        "Calculating bucket sizes\r\n",
        "  Binary sorting into buckets\r\n",
        "  10%\r\n",
        "  20%\r\n",
        "  30%\r\n"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "  40%\r\n",
        "  50%\r\n",
        "  60%\r\n",
        "  70%\r\n",
        "  80%\r\n",
        "  90%\r\n",
        "  100%\r\n",
        "  Binary sorting into buckets time: 00:00:00\r\n",
        "Splitting and merging\r\n",
        "  Splitting and merging time: 00:00:00\r\n",
        "Split 1, merged 6; iterating...\r\n",
        "  Binary sorting into buckets\r\n",
        "  10%\r\n",
        "  20%\r\n",
        "  30%\r\n",
        "  40%\r\n",
        "  50%\r\n",
        "  60%\r\n",
        "  70%\r\n",
        "  80%\r\n",
        "  90%\r\n",
        "  100%\r\n",
        "  Binary sorting into buckets time: 00:00:00\r\n",
        "Splitting and merging\r\n",
        "  Splitting and merging time: 00:00:00\r\n",
        "Split 1, merged 1; iterating...\r\n",
        "  Binary sorting into buckets\r\n",
        "  10%\r\n",
        "  20%\r\n",
        "  30%\r\n",
        "  40%\r\n",
        "  50%\r\n",
        "  60%\r\n",
        "  70%\r\n",
        "  80%\r\n",
        "  90%\r\n",
        "  100%\r\n",
        "  Binary sorting into buckets time: 00:00:00\r\n",
        "Splitting and merging\r\n",
        "  Splitting and merging time: 00:00:00\r\n",
        "Avg bucket size: 31266.6 (target: 46901)\r\n",
        "Converting suffix-array elements to index image\r\n",
        "Allocating ftab, absorbFtab\r\n",
        "Entering Ebwt loop\r\n",
        "Getting block 1 of 8\r\n",
        "  Reserving size (46902) for bucket\r\n",
        "  Calculating Z arrays\r\n",
        "  Calculating Z arrays time: 00:00:00\r\n",
        "  Entering block accumulator loop:\r\n",
        "  10%\r\n",
        "  20%\r\n",
        "  30%\r\n",
        "  40%\r\n",
        "  50%\r\n",
        "  60%\r\n",
        "  70%\r\n",
        "  80%\r\n",
        "  90%\r\n",
        "  100%\r\n",
        "  Block accumulator loop time: 00:00:00\r\n",
        "  Sorting block of length 35684\r\n",
        "  (Using difference cover)\r\n",
        "  Sorting block time: 00:00:00\r\n",
        "Returning block of 35685\r\n",
        "Getting block 2 of 8\r\n",
        "  Reserving size (46902) for bucket\r\n",
        "  Calculating Z arrays\r\n",
        "  Calculating Z arrays time: 00:00:00\r\n",
        "  Entering block accumulator loop:\r\n",
        "  10%\r\n",
        "  20%\r\n",
        "  30%\r\n",
        "  40%\r\n",
        "  50%\r\n",
        "  60%\r\n",
        "  70%\r\n",
        "  80%\r\n",
        "  90%\r\n",
        "  100%\r\n",
        "  Block accumulator loop time: 00:00:00\r\n",
        "  Sorting block of length 14882\r\n",
        "  (Using difference cover)\r\n",
        "  Sorting block time: 00:00:00\r\n",
        "Returning block of 14883\r\n",
        "Getting block 3 of 8\r\n",
        "  Reserving size (46902) for bucket\r\n",
        "  Calculating Z arrays\r\n",
        "  Calculating Z arrays time: 00:00:00\r\n",
        "  Entering block accumulator loop:\r\n",
        "  10%\r\n",
        "  20%\r\n",
        "  30%\r\n",
        "  40%\r\n",
        "  50%\r\n"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "  60%\r\n",
        "  70%\r\n",
        "  80%\r\n",
        "  90%\r\n",
        "  100%\r\n",
        "  Block accumulator loop time: 00:00:00\r\n",
        "  Sorting block of length 43445\r\n",
        "  (Using difference cover)\r\n",
        "  Sorting block time: 00:00:00\r\n",
        "Returning block of 43446\r\n",
        "Getting block 4 of 8\r\n",
        "  Reserving size (46902) for bucket\r\n",
        "  Calculating Z arrays\r\n",
        "  Calculating Z arrays time: 00:00:00\r\n",
        "  Entering block accumulator loop:\r\n",
        "  10%\r\n",
        "  20%\r\n",
        "  30%\r\n",
        "  40%\r\n",
        "  50%\r\n",
        "  60%\r\n",
        "  70%\r\n",
        "  80%\r\n",
        "  90%\r\n",
        "  100%\r\n",
        "  Block accumulator loop time: 00:00:00\r\n",
        "  Sorting block of length 38216\r\n",
        "  (Using difference cover)\r\n",
        "  Sorting block time: 00:00:00\r\n",
        "Returning block of 38217\r\n",
        "Getting block 5 of 8\r\n",
        "  Reserving size (46902) for bucket\r\n",
        "  Calculating Z arrays\r\n",
        "  Calculating Z arrays time: 00:00:00\r\n",
        "  Entering block accumulator loop:\r\n",
        "  10%\r\n",
        "  20%\r\n",
        "  30%\r\n",
        "  40%\r\n",
        "  50%\r\n",
        "  60%\r\n",
        "  70%\r\n",
        "  80%\r\n",
        "  90%\r\n",
        "  100%\r\n",
        "  Block accumulator loop time: 00:00:00\r\n",
        "  Sorting block of length 31757\r\n",
        "  (Using difference cover)\r\n",
        "  Sorting block time: 00:00:00\r\n",
        "Returning block of 31758\r\n",
        "Getting block 6 of 8\r\n",
        "  Reserving size (46902) for bucket\r\n",
        "  Calculating Z arrays\r\n",
        "  Calculating Z arrays time: 00:00:00\r\n",
        "  Entering block accumulator loop:\r\n",
        "  10%\r\n",
        "  20%\r\n",
        "  30%\r\n",
        "  40%\r\n",
        "  50%\r\n",
        "  60%\r\n",
        "  70%\r\n",
        "  80%\r\n",
        "  90%\r\n",
        "  100%\r\n",
        "  Block accumulator loop time: 00:00:00\r\n",
        "  Sorting block of length 23628\r\n",
        "  (Using difference cover)\r\n",
        "  Sorting block time: 00:00:00\r\n",
        "Returning block of 23629\r\n"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "Getting block 7 of 8\r\n",
        "  Reserving size (46902) for bucket\r\n",
        "  Calculating Z arrays\r\n",
        "  Calculating Z arrays time: 00:00:00\r\n",
        "  Entering block accumulator loop:\r\n",
        "  10%\r\n",
        "  20%\r\n",
        "  30%\r\n",
        "  40%\r\n",
        "  50%\r\n",
        "  60%\r\n",
        "  70%\r\n",
        "  80%\r\n",
        "  90%\r\n",
        "  100%\r\n",
        "  Block accumulator loop time: 00:00:00\r\n",
        "  Sorting block of length 41874\r\n",
        "  (Using difference cover)\r\n",
        "  Sorting block time: 00:00:00\r\n",
        "Returning block of 41875\r\n",
        "Getting block 8 of 8\r\n",
        "  Reserving size (46902) for bucket\r\n",
        "  Calculating Z arrays\r\n",
        "  Calculating Z arrays time: 00:00:00\r\n",
        "  Entering block accumulator loop:\r\n",
        "  10%\r\n",
        "  20%\r\n",
        "  30%\r\n",
        "  40%\r\n",
        "  50%\r\n",
        "  60%\r\n",
        "  70%\r\n",
        "  80%\r\n",
        "  90%\r\n",
        "  100%\r\n",
        "  Block accumulator loop time: 00:00:00\r\n",
        "  Sorting block of length 20647\r\n",
        "  (Using difference cover)\r\n",
        "  Sorting block time: 00:00:00\r\n",
        "Returning block of 20648\r\n",
        "Exited Ebwt loop\r\n",
        "fchr[A]: 0\r\n",
        "fchr[C]: 60830\r\n",
        "fchr[G]: 113561\r\n",
        "fchr[T]: 186940\r\n",
        "fchr[$]: 250140\r\n"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "Exiting Ebwt::buildToDisk()\r\n"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "Returning from initFromVector\r\n",
        "Wrote 4358514 bytes to primary EBWT file: GeneSeqs_DB/P1.1.bt2\r\n",
        "Wrote 62540 bytes to secondary EBWT file: GeneSeqs_DB/P1.2.bt2\r\n",
        "Re-opening _in1 and _in2 as input streams\r\n",
        "Returning from Ebwt constructor\r\n",
        "Headers:\r\n",
        "    len: 250140\r\n",
        "    bwtLen: 250141\r\n",
        "    sz: 62535\r\n",
        "    bwtSz: 62536\r\n",
        "    lineRate: 6\r\n",
        "    offRate: 4\r\n",
        "    offMask: 0xfffffff0\r\n",
        "    ftabChars: 10\r\n",
        "    eftabLen: 20\r\n",
        "    eftabSz: 80\r\n",
        "    ftabLen: 1048577\r\n",
        "    ftabSz: 4194308\r\n",
        "    offsLen: 15634\r\n",
        "    offsSz: 62536\r\n",
        "    lineSz: 64\r\n",
        "    sideSz: 64\r\n",
        "    sideBwtSz: 48\r\n",
        "    sideBwtLen: 192\r\n",
        "    numSides: 1303\r\n",
        "    numLines: 1303\r\n",
        "    ebwtTotLen: 83392\r\n",
        "    ebwtTotSz: 83392\r\n",
        "    color: 0\r\n",
        "    reverse: 0\r\n"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "Total time for call to driver() for forward index: 00:00:02\r\n",
        "Reading reference sizes\r\n",
        "Warning: Encountered reference sequence with only gaps\r\n",
        "  Time reading reference sizes: 00:00:00\r\n",
        "Calculating joined length\r\n",
        "Writing header\r\n",
        "Reserving space for joined string\r\n",
        "Joining reference sequences\r\n",
        "  Time to join reference sequences: 00:00:00\r\n",
        "  Time to reverse reference sequence: 00:00:00\r\n",
        "bmax according to bmaxDivN setting: 62535\r\n",
        "Using parameters --bmax 46902 --dcv 1024\r\n",
        "  Doing ahead-of-time memory usage test\r\n",
        "  Passed!  Constructing with these parameters: --bmax 46902 --dcv 1024\r\n",
        "Constructing suffix-array element generator\r\n",
        "Building DifferenceCoverSample\r\n",
        "  Building sPrime\r\n",
        "  Building sPrimeOrder\r\n",
        "  V-Sorting samples\r\n",
        "  V-Sorting samples time: 00:00:00\r\n",
        "  Allocating rank array\r\n",
        "  Ranking v-sort output\r\n",
        "  Ranking v-sort output time: 00:00:00\r\n",
        "  Invoking Larsson-Sadakane on ranks\r\n",
        "  Invoking Larsson-Sadakane on ranks time: 00:00:00\r\n",
        "  Sanity-checking and returning\r\n",
        "Building samples\r\n",
        "Reserving space for 12 sample suffixes\r\n",
        "Generating random suffixes\r\n",
        "QSorting 12 sample offsets, eliminating duplicates\r\n",
        "QSorting sample offsets, eliminating duplicates time: 00:00:00\r\n",
        "Multikey QSorting 12 samples\r\n",
        "  (Using difference cover)\r\n",
        "  Multikey QSorting samples time: 00:00:00\r\n",
        "Calculating bucket sizes\r\n",
        "  Binary sorting into buckets\r\n",
        "  10%\r\n",
        "  20%\r\n",
        "  30%\r\n",
        "  40%\r\n",
        "  50%\r\n",
        "  60%\r\n",
        "  70%\r\n",
        "  80%\r\n",
        "  90%\r\n",
        "  100%\r\n",
        "  Binary sorting into buckets time: 00:00:00\r\n",
        "Splitting and merging\r\n",
        "  Splitting and merging time: 00:00:00\r\n",
        "Split 1, merged 6; iterating...\r\n",
        "  Binary sorting into buckets\r\n",
        "  10%\r\n",
        "  20%\r\n",
        "  30%\r\n",
        "  40%\r\n",
        "  50%\r\n",
        "  60%\r\n",
        "  70%\r\n",
        "  80%\r\n",
        "  90%\r\n",
        "  100%\r\n",
        "  Binary sorting into buckets time: 00:00:00\r\n",
        "Splitting and merging\r\n",
        "  Splitting and merging time: 00:00:00\r\n",
        "Avg bucket size: 31266.6 (target: 46901)\r\n",
        "Converting suffix-array elements to index image\r\n",
        "Allocating ftab, absorbFtab\r\n",
        "Entering Ebwt loop\r\n",
        "Getting block 1 of 8\r\n",
        "  Reserving size (46902) for bucket\r\n",
        "  Calculating Z arrays\r\n",
        "  Calculating Z arrays time: 00:00:00\r\n",
        "  Entering block accumulator loop:\r\n",
        "  10%\r\n",
        "  20%\r\n",
        "  30%\r\n",
        "  40%\r\n",
        "  50%\r\n",
        "  60%\r\n",
        "  70%\r\n",
        "  80%\r\n",
        "  90%\r\n",
        "  100%\r\n",
        "  Block accumulator loop time: 00:00:00\r\n",
        "  Sorting block of length 20273\r\n",
        "  (Using difference cover)\r\n",
        "  Sorting block time: 00:00:00\r\n",
        "Returning block of 20274\r\n",
        "Getting block 2 of 8\r\n",
        "  Reserving size (46902) for bucket\r\n",
        "  Calculating Z arrays\r\n",
        "  Calculating Z arrays time: 00:00:00\r\n",
        "  Entering block accumulator loop:\r\n",
        "  10%\r\n",
        "  20%\r\n",
        "  30%\r\n",
        "  40%\r\n"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "  50%\r\n",
        "  60%\r\n",
        "  70%\r\n",
        "  80%\r\n",
        "  90%\r\n",
        "  100%\r\n",
        "  Block accumulator loop time: 00:00:00\r\n",
        "  Sorting block of length 43929\r\n",
        "  (Using difference cover)\r\n",
        "  Sorting block time: 00:00:00\r\n",
        "Returning block of 43930\r\n",
        "Getting block 3 of 8\r\n",
        "  Reserving size (46902) for bucket\r\n",
        "  Calculating Z arrays\r\n",
        "  Calculating Z arrays time: 00:00:00\r\n",
        "  Entering block accumulator loop:\r\n",
        "  10%\r\n",
        "  20%\r\n",
        "  30%\r\n",
        "  40%\r\n",
        "  50%\r\n",
        "  60%\r\n",
        "  70%\r\n",
        "  80%\r\n",
        "  90%\r\n",
        "  100%\r\n",
        "  Block accumulator loop time: 00:00:00\r\n",
        "  Sorting block of length 45374\r\n",
        "  (Using difference cover)\r\n",
        "  Sorting block time: 00:00:00\r\n",
        "Returning block of 45375\r\n",
        "Getting block 4 of 8\r\n",
        "  Reserving size (46902) for bucket\r\n",
        "  Calculating Z arrays\r\n",
        "  Calculating Z arrays time: 00:00:00\r\n",
        "  Entering block accumulator loop:\r\n",
        "  10%\r\n",
        "  20%\r\n",
        "  30%\r\n",
        "  40%\r\n",
        "  50%\r\n",
        "  60%\r\n",
        "  70%\r\n",
        "  80%\r\n",
        "  90%\r\n",
        "  100%\r\n",
        "  Block accumulator loop time: 00:00:00\r\n",
        "  Sorting block of length 22023\r\n",
        "  (Using difference cover)\r\n",
        "  Sorting block time: 00:00:00\r\n",
        "Returning block of 22024\r\n",
        "Getting block 5 of 8\r\n",
        "  Reserving size (46902) for bucket\r\n",
        "  Calculating Z arrays\r\n",
        "  Calculating Z arrays time: 00:00:00\r\n",
        "  Entering block accumulator loop:\r\n",
        "  10%\r\n",
        "  20%\r\n",
        "  30%\r\n",
        "  40%\r\n",
        "  50%\r\n",
        "  60%\r\n",
        "  70%\r\n",
        "  80%\r\n",
        "  90%\r\n",
        "  100%\r\n",
        "  Block accumulator loop time: 00:00:00\r\n",
        "  Sorting block of length 31539\r\n",
        "  (Using difference cover)\r\n",
        "  Sorting block time: 00:00:00\r\n",
        "Returning block of 31540\r\n",
        "Getting block 6 of 8\r\n",
        "  Reserving size (46902) for bucket\r\n",
        "  Calculating Z arrays\r\n",
        "  Calculating Z arrays time: 00:00:00\r\n",
        "  Entering block accumulator loop:\r\n",
        "  10%\r\n",
        "  20%\r\n",
        "  30%\r\n",
        "  40%\r\n",
        "  50%\r\n",
        "  60%\r\n",
        "  70%\r\n",
        "  80%\r\n",
        "  90%\r\n",
        "  100%\r\n",
        "  Block accumulator loop time: 00:00:00\r\n",
        "  Sorting block of length 30657\r\n",
        "  (Using difference cover)\r\n",
        "  Sorting block time: 00:00:00\r\n",
        "Returning block of 30658\r\n",
        "Getting block 7 of 8\r\n",
        "  Reserving size (46902) for bucket\r\n",
        "  Calculating Z arrays\r\n",
        "  Calculating Z arrays time: 00:00:00\r\n",
        "  Entering block accumulator loop:\r\n",
        "  10%\r\n",
        "  20%\r\n",
        "  30%\r\n",
        "  40%\r\n",
        "  50%\r\n",
        "  60%\r\n",
        "  70%\r\n",
        "  80%\r\n",
        "  90%\r\n",
        "  100%\r\n",
        "  Block accumulator loop time: 00:00:00\r\n",
        "  Sorting block of length 27711\r\n",
        "  (Using difference cover)\r\n",
        "  Sorting block time: 00:00:00\r\n",
        "Returning block of 27712\r\n",
        "Getting block 8 of 8\r\n",
        "  Reserving size (46902) for bucket\r\n",
        "  Calculating Z arrays\r\n",
        "  Calculating Z arrays time: 00:00:00\r\n",
        "  Entering block accumulator loop:\r\n",
        "  10%\r\n",
        "  20%\r\n",
        "  30%\r\n",
        "  40%\r\n",
        "  50%\r\n",
        "  60%\r\n",
        "  70%\r\n",
        "  80%\r\n",
        "  90%\r\n",
        "  100%\r\n",
        "  Block accumulator loop time: 00:00:00\r\n",
        "  Sorting block of length 28627\r\n",
        "  (Using difference cover)\r\n"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "  Sorting block time: 00:00:00\r\n",
        "Returning block of 28628\r\n",
        "Exited Ebwt loop\r\n",
        "fchr[A]: 0\r\n",
        "fchr[C]: 60830\r\n",
        "fchr[G]: 113561\r\n",
        "fchr[T]: 186940\r\n",
        "fchr[$]: 250140\r\n",
        "Exiting Ebwt::buildToDisk()\r\n",
        "Returning from initFromVector\r\n",
        "Wrote 4358514 bytes to primary EBWT file: GeneSeqs_DB/P1.rev.1.bt2\r\n",
        "Wrote 62540 bytes to secondary EBWT file: GeneSeqs_DB/P1.rev.2.bt2\r\n",
        "Re-opening _in1 and _in2 as input streams\r\n",
        "Returning from Ebwt constructor\r\n",
        "Headers:\r\n",
        "    len: 250140\r\n",
        "    bwtLen: 250141\r\n",
        "    sz: 62535\r\n",
        "    bwtSz: 62536\r\n",
        "    lineRate: 6\r\n",
        "    offRate: 4\r\n",
        "    offMask: 0xfffffff0\r\n",
        "    ftabChars: 10\r\n",
        "    eftabLen: 20\r\n",
        "    eftabSz: 80\r\n",
        "    ftabLen: 1048577\r\n",
        "    ftabSz: 4194308\r\n",
        "    offsLen: 15634\r\n",
        "    offsSz: 62536\r\n",
        "    lineSz: 64\r\n",
        "    sideSz: 64\r\n",
        "    sideBwtSz: 48\r\n",
        "    sideBwtLen: 192\r\n",
        "    numSides: 1303\r\n",
        "    numLines: 1303\r\n",
        "    ebwtTotLen: 83392\r\n",
        "    ebwtTotSz: 83392\r\n",
        "    color: 0\r\n",
        "    reverse: 1\r\n",
        "Total time for backward call to driver() for mirror index: 00:00:00\r\n"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "Settings:\r\n",
        "  Output files: \"GeneSeqs_DB/P2.*.bt2\"\r\n",
        "  Line rate: 6 (line is 64 bytes)\r\n",
        "  Lines per side: 1 (side is 64 bytes)\r\n",
        "  Offset rate: 4 (one in 16)\r\n",
        "  FTable chars: 10\r\n",
        "  Strings: unpacked\r\n",
        "  Max bucket size: default\r\n",
        "  Max bucket size, sqrt multiplier: default\r\n",
        "  Max bucket size, len divisor: 4\r\n",
        "  Difference-cover sample period: 1024\r\n",
        "  Endianness: little\r\n",
        "  Actual local endianness: little\r\n",
        "  Sanity checking: disabled\r\n",
        "  Assertions: disabled\r\n",
        "  Random seed: 0\r\n",
        "  Sizeofs: void*:8, int:4, long:8, size_t:8\r\n",
        "Input files DNA, FASTA:\r\n",
        "  GeneSeqs_DB/primer2.fna\r\n",
        "Building a SMALL index\r\n",
        "Reading reference sizes\r\n",
        "Warning: Encountered reference sequence with only gaps\r\n"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "  Time reading reference sizes: 00:00:00\r\n",
        "Calculating joined length\r\n",
        "Writing header\r\n",
        "Reserving space for joined string\r\n",
        "Joining reference sequences\r\n",
        "  Time to join reference sequences: 00:00:00\r\n",
        "bmax according to bmaxDivN setting: 62535\r\n",
        "Using parameters --bmax 46902 --dcv 1024\r\n",
        "  Doing ahead-of-time memory usage test\r\n",
        "  Passed!  Constructing with these parameters: --bmax 46902 --dcv 1024\r\n",
        "Constructing suffix-array element generator\r\n",
        "Building DifferenceCoverSample\r\n",
        "  Building sPrime\r\n",
        "  Building sPrimeOrder\r\n",
        "  V-Sorting samples\r\n",
        "  V-Sorting samples time: 00:00:00\r\n",
        "  Allocating rank array\r\n",
        "  Ranking v-sort output\r\n",
        "  Ranking v-sort output time: 00:00:00\r\n",
        "  Invoking Larsson-Sadakane on ranks\r\n",
        "  Invoking Larsson-Sadakane on ranks time: 00:00:00\r\n",
        "  Sanity-checking and returning\r\n",
        "Building samples\r\n",
        "Reserving space for 12 sample suffixes\r\n",
        "Generating random suffixes\r\n",
        "QSorting 12 sample offsets, eliminating duplicates\r\n",
        "QSorting sample offsets, eliminating duplicates time: 00:00:00\r\n",
        "Multikey QSorting 12 samples\r\n",
        "  (Using difference cover)\r\n",
        "  Multikey QSorting samples time: 00:00:00\r\n",
        "Calculating bucket sizes\r\n",
        "  Binary sorting into buckets\r\n",
        "  10%\r\n",
        "  20%\r\n",
        "  30%\r\n",
        "  40%\r\n",
        "  50%\r\n",
        "  60%\r\n",
        "  70%\r\n",
        "  80%\r\n",
        "  90%\r\n",
        "  100%\r\n",
        "  Binary sorting into buckets time: 00:00:00\r\n",
        "Splitting and merging\r\n",
        "  Splitting and merging time: 00:00:00\r\n",
        "Split 2, merged 5; iterating...\r\n",
        "  Binary sorting into buckets\r\n",
        "  10%\r\n",
        "  20%\r\n",
        "  30%\r\n",
        "  40%\r\n",
        "  50%\r\n",
        "  60%\r\n",
        "  70%\r\n",
        "  80%\r\n",
        "  90%\r\n",
        "  100%\r\n",
        "  Binary sorting into buckets time: 00:00:00\r\n",
        "Splitting and merging\r\n",
        "  Splitting and merging time: 00:00:00\r\n",
        "Split 1, merged 2; iterating...\r\n",
        "  Binary sorting into buckets\r\n",
        "  10%\r\n",
        "  20%\r\n",
        "  30%\r\n",
        "  40%\r\n",
        "  50%\r\n",
        "  60%\r\n",
        "  70%\r\n"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "  80%\r\n",
        "  90%\r\n",
        "  100%\r\n",
        "  Binary sorting into buckets time: 00:00:00\r\n",
        "Splitting and merging\r\n",
        "  Splitting and merging time: 00:00:00\r\n",
        "Avg bucket size: 31266.6 (target: 46901)\r\n",
        "Converting suffix-array elements to index image\r\n",
        "Allocating ftab, absorbFtab\r\n",
        "Entering Ebwt loop\r\n",
        "Getting block 1 of 8\r\n",
        "  Reserving size (46902) for bucket\r\n",
        "  Calculating Z arrays\r\n",
        "  Calculating Z arrays time: 00:00:00\r\n",
        "  Entering block accumulator loop:\r\n",
        "  10%\r\n",
        "  20%\r\n",
        "  30%\r\n",
        "  40%\r\n",
        "  50%\r\n",
        "  60%\r\n",
        "  70%\r\n",
        "  80%\r\n",
        "  90%\r\n",
        "  100%\r\n",
        "  Block accumulator loop time: 00:00:00\r\n",
        "  Sorting block of length 24534\r\n",
        "  (Using difference cover)\r\n",
        "  Sorting block time: 00:00:00\r\n",
        "Returning block of 24535\r\n",
        "Getting block 2 of 8\r\n",
        "  Reserving size (46902) for bucket\r\n",
        "  Calculating Z arrays\r\n",
        "  Calculating Z arrays time: 00:00:00\r\n",
        "  Entering block accumulator loop:\r\n",
        "  10%\r\n",
        "  20%\r\n",
        "  30%\r\n",
        "  40%\r\n",
        "  50%\r\n",
        "  60%\r\n",
        "  70%\r\n",
        "  80%\r\n",
        "  90%\r\n",
        "  100%\r\n",
        "  Block accumulator loop time: 00:00:00\r\n",
        "  Sorting block of length 44650\r\n",
        "  (Using difference cover)\r\n",
        "  Sorting block time: 00:00:00\r\n",
        "Returning block of 44651\r\n",
        "Getting block 3 of 8\r\n",
        "  Reserving size (46902) for bucket\r\n",
        "  Calculating Z arrays\r\n",
        "  Calculating Z arrays time: 00:00:00\r\n",
        "  Entering block accumulator loop:\r\n",
        "  10%\r\n",
        "  20%\r\n",
        "  30%\r\n",
        "  40%\r\n",
        "  50%\r\n",
        "  60%\r\n",
        "  70%\r\n",
        "  80%\r\n",
        "  90%\r\n",
        "  100%\r\n",
        "  Block accumulator loop time: 00:00:00\r\n",
        "  Sorting block of length 39842\r\n",
        "  (Using difference cover)\r\n",
        "  Sorting block time: 00:00:00\r\n",
        "Returning block of 39843\r\n",
        "Getting block 4 of 8\r\n",
        "  Reserving size (46902) for bucket\r\n",
        "  Calculating Z arrays\r\n",
        "  Calculating Z arrays time: 00:00:00\r\n",
        "  Entering block accumulator loop:\r\n",
        "  10%\r\n",
        "  20%\r\n",
        "  30%\r\n",
        "  40%\r\n",
        "  50%\r\n",
        "  60%\r\n",
        "  70%\r\n",
        "  80%\r\n",
        "  90%\r\n",
        "  100%\r\n",
        "  Block accumulator loop time: 00:00:00\r\n",
        "  Sorting block of length 13750\r\n",
        "  (Using difference cover)\r\n",
        "  Sorting block time: 00:00:00\r\n",
        "Returning block of 13751\r\n",
        "Getting block 5 of 8\r\n",
        "  Reserving size (46902) for bucket\r\n",
        "  Calculating Z arrays\r\n",
        "  Calculating Z arrays time: 00:00:00\r\n",
        "  Entering block accumulator loop:\r\n",
        "  10%\r\n",
        "  20%\r\n",
        "  30%\r\n",
        "  40%\r\n",
        "  50%\r\n",
        "  60%\r\n",
        "  70%\r\n",
        "  80%\r\n",
        "  90%\r\n",
        "  100%\r\n",
        "  Block accumulator loop time: 00:00:00\r\n",
        "  Sorting block of length 39483\r\n",
        "  (Using difference cover)\r\n",
        "  Sorting block time: 00:00:00\r\n",
        "Returning block of 39484\r\n",
        "Getting block 6 of 8\r\n",
        "  Reserving size (46902) for bucket\r\n",
        "  Calculating Z arrays\r\n",
        "  Calculating Z arrays time: 00:00:00\r\n",
        "  Entering block accumulator loop:\r\n",
        "  10%\r\n",
        "  20%\r\n",
        "  30%\r\n",
        "  40%\r\n",
        "  50%\r\n",
        "  60%\r\n",
        "  70%\r\n",
        "  80%\r\n",
        "  90%\r\n",
        "  100%\r\n",
        "  Block accumulator loop time: 00:00:00\r\n",
        "  Sorting block of length 31985\r\n",
        "  (Using difference cover)\r\n",
        "  Sorting block time: 00:00:00\r\n",
        "Returning block of 31986\r\n",
        "Getting block 7 of 8\r\n",
        "  Reserving size (46902) for bucket\r\n",
        "  Calculating Z arrays\r\n",
        "  Calculating Z arrays time: 00:00:00\r\n",
        "  Entering block accumulator loop:\r\n",
        "  10%\r\n",
        "  20%\r\n"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "  30%\r\n",
        "  40%\r\n",
        "  50%\r\n",
        "  60%\r\n",
        "  70%\r\n",
        "  80%\r\n",
        "  90%\r\n",
        "  100%\r\n",
        "  Block accumulator loop time: 00:00:00\r\n",
        "  Sorting block of length 31741\r\n",
        "  (Using difference cover)\r\n",
        "  Sorting block time: 00:00:00\r\n",
        "Returning block of 31742\r\n",
        "Getting block 8 of 8\r\n",
        "  Reserving size (46902) for bucket\r\n",
        "  Calculating Z arrays\r\n",
        "  Calculating Z arrays time: 00:00:00\r\n",
        "  Entering block accumulator loop:\r\n",
        "  10%\r\n",
        "  20%\r\n",
        "  30%\r\n",
        "  40%\r\n",
        "  50%\r\n",
        "  60%\r\n",
        "  70%\r\n",
        "  80%\r\n",
        "  90%\r\n",
        "  100%\r\n",
        "  Block accumulator loop time: 00:00:00\r\n",
        "  Sorting block of length 24148\r\n",
        "  (Using difference cover)\r\n",
        "  Sorting block time: 00:00:00\r\n",
        "Returning block of 24149\r\n",
        "Exited Ebwt loop\r\n",
        "fchr[A]: 0\r\n",
        "fchr[C]: 65555\r\n",
        "fchr[G]: 128334\r\n",
        "fchr[T]: 173953\r\n",
        "fchr[$]: 250140\r\n",
        "Exiting Ebwt::buildToDisk()\r\n",
        "Returning from initFromVector\r\n",
        "Wrote 4358514 bytes to primary EBWT file: GeneSeqs_DB/P2.1.bt2\r\n"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "Wrote 62540 bytes to secondary EBWT file: GeneSeqs_DB/P2.2.bt2\r\n",
        "Re-opening _in1 and _in2 as input streams\r\n",
        "Returning from Ebwt constructor\r\n",
        "Headers:\r\n",
        "    len: 250140\r\n",
        "    bwtLen: 250141\r\n",
        "    sz: 62535\r\n",
        "    bwtSz: 62536\r\n",
        "    lineRate: 6\r\n",
        "    offRate: 4\r\n",
        "    offMask: 0xfffffff0\r\n",
        "    ftabChars: 10\r\n",
        "    eftabLen: 20\r\n",
        "    eftabSz: 80\r\n",
        "    ftabLen: 1048577\r\n",
        "    ftabSz: 4194308\r\n",
        "    offsLen: 15634\r\n",
        "    offsSz: 62536\r\n",
        "    lineSz: 64\r\n",
        "    sideSz: 64\r\n",
        "    sideBwtSz: 48\r\n",
        "    sideBwtLen: 192\r\n",
        "    numSides: 1303\r\n",
        "    numLines: 1303\r\n",
        "    ebwtTotLen: 83392\r\n",
        "    ebwtTotSz: 83392\r\n",
        "    color: 0\r\n",
        "    reverse: 0\r\n",
        "Total time for call to driver() for forward index: 00:00:00\r\n",
        "Reading reference sizes\r\n",
        "Warning: Encountered reference sequence with only gaps\r\n",
        "  Time reading reference sizes: 00:00:00\r\n",
        "Calculating joined length\r\n",
        "Writing header\r\n",
        "Reserving space for joined string\r\n",
        "Joining reference sequences\r\n",
        "  Time to join reference sequences: 00:00:00\r\n",
        "  Time to reverse reference sequence: 00:00:00\r\n",
        "bmax according to bmaxDivN setting: 62535\r\n",
        "Using parameters --bmax 46902 --dcv 1024\r\n",
        "  Doing ahead-of-time memory usage test\r\n",
        "  Passed!  Constructing with these parameters: --bmax 46902 --dcv 1024\r\n",
        "Constructing suffix-array element generator\r\n",
        "Building DifferenceCoverSample\r\n",
        "  Building sPrime\r\n",
        "  Building sPrimeOrder\r\n",
        "  V-Sorting samples\r\n",
        "  V-Sorting samples time: 00:00:00\r\n",
        "  Allocating rank array\r\n",
        "  Ranking v-sort output\r\n",
        "  Ranking v-sort output time: 00:00:00\r\n",
        "  Invoking Larsson-Sadakane on ranks\r\n",
        "  Invoking Larsson-Sadakane on ranks time: 00:00:00\r\n",
        "  Sanity-checking and returning\r\n",
        "Building samples\r\n",
        "Reserving space for 12 sample suffixes\r\n",
        "Generating random suffixes\r\n",
        "QSorting 12 sample offsets, eliminating duplicates\r\n",
        "QSorting sample offsets, eliminating duplicates time: 00:00:00\r\n",
        "Multikey QSorting 12 samples\r\n",
        "  (Using difference cover)\r\n",
        "  Multikey QSorting samples time: 00:00:00\r\n",
        "Calculating bucket sizes\r\n",
        "  Binary sorting into buckets\r\n",
        "  10%\r\n",
        "  20%\r\n",
        "  30%\r\n",
        "  40%\r\n",
        "  50%\r\n",
        "  60%\r\n",
        "  70%\r\n",
        "  80%\r\n",
        "  90%\r\n",
        "  100%\r\n",
        "  Binary sorting into buckets time: 00:00:00\r\n",
        "Splitting and merging\r\n",
        "  Splitting and merging time: 00:00:00\r\n",
        "Avg bucket size: 35733.4 (target: 46901)\r\n",
        "Converting suffix-array elements to index image\r\n",
        "Allocating ftab, absorbFtab\r\n",
        "Entering Ebwt loop\r\n",
        "Getting block 1 of 7\r\n",
        "  Reserving size (46902) for bucket\r\n",
        "  Calculating Z arrays\r\n",
        "  Calculating Z arrays time: 00:00:00\r\n",
        "  Entering block accumulator loop:\r\n",
        "  10%\r\n",
        "  20%\r\n",
        "  30%\r\n",
        "  40%\r\n",
        "  50%\r\n",
        "  60%\r\n",
        "  70%\r\n",
        "  80%\r\n",
        "  90%\r\n",
        "  100%\r\n",
        "  Block accumulator loop time: 00:00:00\r\n",
        "  Sorting block of length 33096\r\n",
        "  (Using difference cover)\r\n",
        "  Sorting block time: 00:00:00\r\n",
        "Returning block of 33097\r\n",
        "Getting block 2 of 7\r\n",
        "  Reserving size (46902) for bucket\r\n",
        "  Calculating Z arrays\r\n",
        "  Calculating Z arrays time: 00:00:00\r\n",
        "  Entering block accumulator loop:\r\n",
        "  10%\r\n",
        "  20%\r\n",
        "  30%\r\n",
        "  40%\r\n",
        "  50%\r\n",
        "  60%\r\n",
        "  70%\r\n",
        "  80%\r\n",
        "  90%\r\n",
        "  100%\r\n",
        "  Block accumulator loop time: 00:00:00\r\n",
        "  Sorting block of length 46556\r\n",
        "  (Using difference cover)\r\n"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "  Sorting block time: 00:00:00\r\n",
        "Returning block of 46557\r\n",
        "Getting block 3 of 7\r\n",
        "  Reserving size (46902) for bucket\r\n",
        "  Calculating Z arrays\r\n",
        "  Calculating Z arrays time: 00:00:00\r\n",
        "  Entering block accumulator loop:\r\n",
        "  10%\r\n",
        "  20%\r\n",
        "  30%\r\n",
        "  40%\r\n",
        "  50%\r\n",
        "  60%\r\n",
        "  70%\r\n",
        "  80%\r\n",
        "  90%\r\n",
        "  100%\r\n",
        "  Block accumulator loop time: 00:00:00\r\n",
        "  Sorting block of length 30474\r\n",
        "  (Using difference cover)\r\n",
        "  Sorting block time: 00:00:00\r\n",
        "Returning block of 30475\r\n",
        "Getting block 4 of 7\r\n",
        "  Reserving size (46902) for bucket\r\n",
        "  Calculating Z arrays\r\n",
        "  Calculating Z arrays time: 00:00:00\r\n",
        "  Entering block accumulator loop:\r\n",
        "  10%\r\n",
        "  20%\r\n",
        "  30%\r\n",
        "  40%\r\n",
        "  50%\r\n",
        "  60%\r\n",
        "  70%\r\n",
        "  80%\r\n",
        "  90%\r\n",
        "  100%\r\n",
        "  Block accumulator loop time: 00:00:00\r\n",
        "  Sorting block of length 42578\r\n",
        "  (Using difference cover)\r\n",
        "  Sorting block time: 00:00:00\r\n",
        "Returning block of 42579\r\n",
        "Getting block 5 of 7\r\n",
        "  Reserving size (46902) for bucket\r\n",
        "  Calculating Z arrays\r\n",
        "  Calculating Z arrays time: 00:00:00\r\n",
        "  Entering block accumulator loop:\r\n",
        "  10%\r\n",
        "  20%\r\n",
        "  30%\r\n",
        "  40%\r\n",
        "  50%\r\n",
        "  60%\r\n",
        "  70%\r\n",
        "  80%\r\n",
        "  90%\r\n",
        "  100%\r\n",
        "  Block accumulator loop time: 00:00:00\r\n",
        "  Sorting block of length 37964\r\n",
        "  (Using difference cover)\r\n",
        "  Sorting block time: 00:00:00\r\n",
        "Returning block of 37965\r\n",
        "Getting block 6 of 7\r\n",
        "  Reserving size (46902) for bucket\r\n",
        "  Calculating Z arrays\r\n",
        "  Calculating Z arrays time: 00:00:00\r\n",
        "  Entering block accumulator loop:\r\n",
        "  10%\r\n",
        "  20%\r\n",
        "  30%\r\n",
        "  40%\r\n",
        "  50%\r\n",
        "  60%\r\n",
        "  70%\r\n",
        "  80%\r\n",
        "  90%\r\n",
        "  100%\r\n",
        "  Block accumulator loop time: 00:00:00\r\n",
        "  Sorting block of length 31324\r\n",
        "  (Using difference cover)\r\n",
        "  Sorting block time: 00:00:00\r\n",
        "Returning block of 31325\r\n",
        "Getting block 7 of 7\r\n",
        "  Reserving size (46902) for bucket\r\n",
        "  Calculating Z arrays\r\n",
        "  Calculating Z arrays time: 00:00:00\r\n",
        "  Entering block accumulator loop:\r\n",
        "  10%\r\n",
        "  20%\r\n",
        "  30%\r\n",
        "  40%\r\n",
        "  50%\r\n",
        "  60%\r\n",
        "  70%\r\n",
        "  80%\r\n",
        "  90%\r\n",
        "  100%\r\n",
        "  Block accumulator loop time: 00:00:00\r\n",
        "  Sorting block of length 28142\r\n",
        "  (Using difference cover)\r\n",
        "  Sorting block time: 00:00:00\r\n",
        "Returning block of 28143\r\n",
        "Exited Ebwt loop\r\n",
        "fchr[A]: 0\r\n",
        "fchr[C]: 65555\r\n",
        "fchr[G]: 128334\r\n",
        "fchr[T]: 173953\r\n",
        "fchr[$]: 250140\r\n"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "Exiting Ebwt::buildToDisk()\r\n",
        "Returning from initFromVector\r\n",
        "Wrote 4358514 bytes to primary EBWT file: GeneSeqs_DB/P2.rev.1.bt2\r\n",
        "Wrote 62540 bytes to secondary EBWT file: GeneSeqs_DB/P2.rev.2.bt2\r\n",
        "Re-opening _in1 and _in2 as input streams\r\n",
        "Returning from Ebwt constructor\r\n",
        "Headers:\r\n",
        "    len: 250140\r\n",
        "    bwtLen: 250141\r\n",
        "    sz: 62535\r\n",
        "    bwtSz: 62536\r\n",
        "    lineRate: 6\r\n",
        "    offRate: 4\r\n",
        "    offMask: 0xfffffff0\r\n",
        "    ftabChars: 10\r\n",
        "    eftabLen: 20\r\n",
        "    eftabSz: 80\r\n",
        "    ftabLen: 1048577\r\n",
        "    ftabSz: 4194308\r\n",
        "    offsLen: 15634\r\n",
        "    offsSz: 62536\r\n",
        "    lineSz: 64\r\n",
        "    sideSz: 64\r\n",
        "    sideBwtSz: 48\r\n",
        "    sideBwtLen: 192\r\n",
        "    numSides: 1303\r\n",
        "    numLines: 1303\r\n",
        "    ebwtTotLen: 83392\r\n",
        "    ebwtTotSz: 83392\r\n",
        "    color: 0\r\n",
        "    reverse: 1\r\n",
        "Total time for backward call to driver() for mirror index: 00:00:00\r\n"
       ]
      }
     ],
     "prompt_number": 10
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "For bowtie2, I'm using the following parameters\n",
      "\n",
      "    -k 1 Allows only one match\n",
      "    -p 2 Number of threads. I'm running this on my laptop, so just using 2 threads\n",
      "    --very-sensitive-local Uses local alignment\n",
      "    -x Reference database\n",
      "    -1 and -2 R1 and R2 of the reads\n",
      "    \n",
      "    \n",
      "    "
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#Create folder\n",
      "!mkdir mapped_reads\n",
      "\n",
      "\n",
      "#Map the In reads to P1\n",
      "!bowtie2 -k 1 -p 2 --very-sensitive-local -x GeneSeqs_DB/P1 -1 seqs/Salmonella-input_S3_L001_R1_001.fastq.gz -2 seqs/Salmonella-input_S3_L001_R2_001.fastq.gz -S mapped_reads/input_P1.sam\n",
      "#Map the In reads to P2\n",
      "!bowtie2 -k 1 -p 2 --very-sensitive-local -x GeneSeqs_DB/P2 -1 seqs/Salmonella-input_S3_L001_R1_001.fastq.gz -2 seqs/Salmonella-input_S3_L001_R2_001.fastq.gz -S mapped_reads/input_P2.sam\n",
      "\n",
      "#Map the Out reads to P1\n",
      "!bowtie2 -k 1 -p 2 --very-sensitive-local -x GeneSeqs_DB/P1 -1 seqs/Salmonella-pi_S4_L001_R1_001.fastq.gz -2 seqs/Salmonella-pi_S4_L001_R2_001.fastq.gz -S mapped_reads/pi_P1.sam\n",
      "#Map the Out reads to P2\n",
      "!bowtie2 -k 1 -p 2 --very-sensitive-local -x GeneSeqs_DB/P2 -1 seqs/Salmonella-pi_S4_L001_R1_001.fastq.gz -2 seqs/Salmonella-pi_S4_L001_R2_001.fastq.gz -S mapped_reads/pi_P2.sam\n",
      "\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "mkdir: mapped_reads: File exists\r\n"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "5318803 reads; of these:\r\n",
        "  5318803 (100.00%) were paired; of these:\r\n",
        "    5242061 (98.56%) aligned concordantly 0 times\r\n",
        "    76742 (1.44%) aligned concordantly exactly 1 time\r\n",
        "    0 (0.00%) aligned concordantly >1 times\r\n",
        "    ----\r\n",
        "    5242061 pairs aligned concordantly 0 times; of these:\r\n",
        "      14655 (0.28%) aligned discordantly 1 time\r\n"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "    ----\r\n",
        "    5227406 pairs aligned 0 times concordantly or discordantly; of these:\r\n",
        "      10454812 mates make up the pairs; of these:\r\n",
        "        10029999 (95.94%) aligned 0 times\r\n",
        "        424665 (4.06%) aligned exactly 1 time\r\n",
        "        148 (0.00%) aligned >1 times\r\n",
        "5.71% overall alignment rate\r\n"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "5318803 reads; of these:\r\n",
        "  5318803 (100.00%) were paired; of these:\r\n",
        "    5249664 (98.70%) aligned concordantly 0 times\r\n",
        "    69139 (1.30%) aligned concordantly exactly 1 time\r\n",
        "    0 (0.00%) aligned concordantly >1 times\r\n",
        "    ----\r\n",
        "    5249664 pairs aligned concordantly 0 times; of these:\r\n",
        "      14871 (0.28%) aligned discordantly 1 time\r\n",
        "    ----\r\n",
        "    5234793 pairs aligned 0 times concordantly or discordantly; of these:\r\n",
        "      10469586 mates make up the pairs; of these:\r\n",
        "        10061833 (96.11%) aligned 0 times\r\n",
        "        407599 (3.89%) aligned exactly 1 time\r\n",
        "        154 (0.00%) aligned >1 times\r\n",
        "5.41% overall alignment rate\r\n"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "4426039 reads; of these:\r\n",
        "  4426039 (100.00%) were paired; of these:\r\n",
        "    4367863 (98.69%) aligned concordantly 0 times\r\n",
        "    58176 (1.31%) aligned concordantly exactly 1 time\r\n",
        "    0 (0.00%) aligned concordantly >1 times\r\n",
        "    ----\r\n",
        "    4367863 pairs aligned concordantly 0 times; of these:\r\n",
        "      11178 (0.26%) aligned discordantly 1 time\r\n",
        "    ----\r\n",
        "    4356685 pairs aligned 0 times concordantly or discordantly; of these:\r\n",
        "      8713370 mates make up the pairs; of these:\r\n",
        "        8352626 (95.86%) aligned 0 times\r\n",
        "        360653 (4.14%) aligned exactly 1 time\r\n",
        "        91 (0.00%) aligned >1 times\r\n",
        "5.64% overall alignment rate\r\n"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "4426039 reads; of these:\r\n",
        "  4426039 (100.00%) were paired; of these:\r\n",
        "    4373967 (98.82%) aligned concordantly 0 times\r\n",
        "    52072 (1.18%) aligned concordantly exactly 1 time\r\n",
        "    0 (0.00%) aligned concordantly >1 times\r\n",
        "    ----\r\n",
        "    4373967 pairs aligned concordantly 0 times; of these:\r\n",
        "      11217 (0.26%) aligned discordantly 1 time\r\n",
        "    ----\r\n",
        "    4362750 pairs aligned 0 times concordantly or discordantly; of these:\r\n",
        "      8725500 mates make up the pairs; of these:\r\n",
        "        8377564 (96.01%) aligned 0 times\r\n",
        "        347834 (3.99%) aligned exactly 1 time\r\n",
        "        102 (0.00%) aligned >1 times\r\n",
        "5.36% overall alignment rate\r\n"
       ]
      }
     ],
     "prompt_number": 11
    },
    {
     "cell_type": "heading",
     "level": 2,
     "metadata": {},
     "source": [
      "Looking at the output files (SAM format)"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "First I will create the bam files (a compressed version of SAM). This will help me to save hard drive space. The SAM files created by Bowtie2, have ~2GB each, while the BAM version is ~500Mb. "
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#Create bam files\n",
      "!samtools view -b -o mapped_reads/input_P1.bam mapped_reads/input_P1.sam\n",
      "!samtools view -b -o mapped_reads/input_P2.bam mapped_reads/input_P2.sam\n",
      "!samtools view -b -o mapped_reads/pi_P1.bam mapped_reads/pi_P1.sam\n",
      "!samtools view -b -o mapped_reads/pi_P2.bam mapped_reads/pi_P2.sam"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 13
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#Delete sam files to free space\n",
      "!rm mapped_reads/input_P1.sam\n",
      "!rm mapped_reads/input_P2.sam\n",
      "!rm mapped_reads/pi_P1.sam\n",
      "!rm mapped_reads/pi_P2.sam"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 14
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#Create bam files only with the reads that had matches, sort them and index.\n",
      "!samtools view -b -F 4 -o mapped_reads/aln.input_P1.bam mapped_reads/input_P1.bam\n",
      "!samtools sort -T temp -o mapped_reads/sorted.aln.input_P1.bam mapped_reads/aln.input_P1.bam\n",
      "!samtools index mapped_reads/sorted.aln.input_P1.bam\n",
      "\n",
      "!samtools view -b -F 4 -o mapped_reads/aln.input_P2.bam mapped_reads/input_P2.bam\n",
      "!samtools sort -T temp -o mapped_reads/sorted.aln.input_P2.bam mapped_reads/aln.input_P2.bam\n",
      "!samtools index mapped_reads/sorted.aln.input_P2.bam\n",
      "\n",
      "!samtools view -b -F 4 -o mapped_reads/aln.pi_P1.bam mapped_reads/pi_P1.bam\n",
      "!samtools sort -T temp -o mapped_reads/sorted.aln.pi_P1.bam mapped_reads/aln.pi_P1.bam\n",
      "!samtools index mapped_reads/sorted.aln.pi_P1.bam\n",
      "\n",
      "!samtools view -b -F 4 -o mapped_reads/aln.pi_P2.bam mapped_reads/pi_P2.bam\n",
      "!samtools sort -T temp -o mapped_reads/sorted.aln.pi_P2.bam mapped_reads/aln.pi_P2.bam\n",
      "!samtools index mapped_reads/sorted.aln.pi_P2.bam\n",
      "\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 15
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#Get the number of reads that mapped to each reference, using samtools\n",
      "!samtools idxstats mapped_reads/sorted.aln.input_P1.bam > mapped_reads/stats.input_P1\n",
      "!samtools idxstats mapped_reads/sorted.aln.input_P2.bam > mapped_reads/stats.input_P2\n",
      "!samtools idxstats mapped_reads/sorted.aln.pi_P1.bam > mapped_reads/stats.pi_P1\n",
      "!samtools idxstats mapped_reads/sorted.aln.pi_P2.bam > mapped_reads/stats.pi_P2"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 18
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#Read each file, and store the values in a dictionary\n",
      "count_input_P1 = defaultdict()\n",
      "count_input_P2 = defaultdict()\n",
      "count_pi_P1 = defaultdict()\n",
      "count_pi_P2 = defaultdict()\n",
      "\n",
      "with open(\"mapped_reads/stats.input_P1\") as fp:\n",
      "    for line in fp:\n",
      "        line = line.rstrip()\n",
      "        gene_id, length, count, not_mapped = line.split(\"\\t\")\n",
      "        count_input_P1[gene_id] = int(count)\n",
      "\n",
      "with open(\"mapped_reads/stats.input_P2\") as fp:\n",
      "    for line in fp:\n",
      "        line = line.rstrip()\n",
      "        gene_id, length, count, not_mapped = line.split(\"\\t\")\n",
      "        count_input_P2[gene_id] = int(count)\n",
      "        \n",
      "with open(\"mapped_reads/stats.pi_P1\") as fp:\n",
      "    for line in fp:\n",
      "        line = line.rstrip()\n",
      "        gene_id, length, count, not_mapped = line.split(\"\\t\")\n",
      "        count_pi_P1[gene_id] = int(count)\n",
      "        \n",
      "with open(\"mapped_reads/stats.pi_P2\") as fp:\n",
      "    for line in fp:\n",
      "        line = line.rstrip()\n",
      "        gene_id, length, count, not_mapped = line.split(\"\\t\")\n",
      "        count_pi_P2[gene_id] = int(count)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 4
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#Make a big table output, using the original table as reference\n",
      "#I will read that table again, just to keep the original order\n",
      "\n",
      "table_with_counts = open(\"GeneTable_WithCounts.txt\", 'w')\n",
      "table_with_ra = open(\"GeneTable_WithRelativeAbundance.txt\", 'w')\n",
      "\n",
      "total_input_P1 = sum(count_input_P1.values())\n",
      "total_input_P2 = sum(count_input_P2.values())\n",
      "total_pi_P1 = sum(count_pi_P1.values())\n",
      "total_pi_P2 = sum(count_pi_P2.values())\n",
      "\n",
      "with open(\"GeneSeqTable.txt\") as fp:\n",
      "    for line in fp:\n",
      "        line = line.rstrip()\n",
      "        if line.strip():\n",
      "            running_no, target_gene, gene_LT2, desc, primer1, primer2 = line.split(\"\\t\")\n",
      "            \n",
      "            #I will use the running no as the ID to store in the dictionary. I need to check also\n",
      "            #that is not present\n",
      "            \n",
      "            #Get the values\n",
      "            if running_no in count_input_P1:\n",
      "                input_p1 = count_input_P1[running_no]\n",
      "                ra_input_p1 = input_p1 / float(total_input_P1) * 100\n",
      "            else:\n",
      "                input_p1 = 0\n",
      "                ra_input_p1 = 0\n",
      "            \n",
      "            if running_no in count_input_P2:\n",
      "                input_p2 = count_input_P2[running_no]\n",
      "                ra_input_p2 = input_p2 / float(total_input_P2) * 100\n",
      "            else:\n",
      "                input_p2 = 0\n",
      "                ra_input_p2 = 0\n",
      "                \n",
      "            if running_no in count_pi_P1:\n",
      "                pi_p1 = count_pi_P1[running_no]\n",
      "                ra_pi_p1 = pi_p1 / float(total_pi_P1) * 100\n",
      "            else:\n",
      "                pi_p1 = 0\n",
      "                ra_pi_p1 = 0\n",
      "            \n",
      "            if running_no in count_pi_P2:\n",
      "                pi_p2 = count_pi_P2[running_no]\n",
      "                ra_pi_p2 = pi_p2 / float(total_pi_P2) * 100\n",
      "            else:\n",
      "                pi_p2 = 0\n",
      "                ra_pi_p2 = 0\n",
      "                \n",
      "            #Print everything to the new file\n",
      "            out_values = [input_p1, input_p2, pi_p1, pi_p2]\n",
      "            #out_relative_abundace = [ra_input_p1, ra_input_p2, ra_pi_p1, ra_pi_p2]\n",
      "            \n",
      "            table_with_counts.write(\"\\t\".join(map(str, out_values)) + \"\\t\" + line + \"\\n\")\n",
      "            \n",
      "            table_with_ra.write(\"%.3f\\t%.3f\\t%.3f\\t%.3f\\t\" % (ra_input_p1, ra_input_p2, ra_pi_p1, ra_pi_p2))\n",
      "            table_with_ra.write(line + \"\\n\")\n",
      "            \n",
      "table_with_counts.close()\n",
      "table_with_ra.close()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 5
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "print total_input_P1, total_input_P2, total_pi_P1, total_pi_P2"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "607607 575773 499452 474514\n"
       ]
      }
     ],
     "prompt_number": 6
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}